### PR TITLE
check presence of all bundles before deprecate

### PR DESCRIPTION
**Pull request by ankitathomas**
Allow deprecatetruncate to handle multiple bundles on the same upgrade chain. Only the most latest bundle in the chain will be marked as deprecated, the remaining will be truncated with the rest of the upgrade chain older than the deprecated bundle.

### CODE REVIEW

File: pkg/sqlite/deprecate.go [hunk 1]
```diff
@@ -35,10 +38,35 @@ func (d *BundleDeprecator) Deprecate() error {
...
+	type nameVersion struct {
+		name    string
+		version string
+	}
+	bundleInfo := make(map[string]nameVersion)
+	// Check if all bundlepaths are valid
 	for _, bundlePath := range d.bundles {
-		if err := d.store.DeprecateBundle(bundlePath); err != nil {
-			if !errors.Is(err, registry.ErrBundleImageNotInDatabase) && !errors.Is(err, registry.ErrRemovingDefaultChannelDuringDeprecation) {
+		name, version, err := d.querier.GetBundleNameAndVersionForImage(context.TODO(), bundlePath)
```
Comment (estroz) [inline pkg/sqlite/deprecate.go:48]: Is `bundlePath` supposed to be a globally unique image tag, or can it be a directory path?

Comment (ankitathomas) [reply pkg/sqlite/deprecate.go]: It's a globally unique tag.

File: pkg/lib/registry/registry.go [hunk 2]
```diff
@@ -342,9 +342,45 @@ func (r RegistryUpdater) DeprecateFromRegistry(request DeprecateFromRegistryRequ
...
+	dbQuerier := sqlite.NewSQLLiteQuerierFromDb(db)
+	registryBundles, err := dbQuerier.ListBundles(context.TODO())
+	if err != nil {
+		if !request.Permissive {
+			return fmt.Errorf("cannot verify bundle presence in registry: %v", err)
+		}
+		toDeprecate = request.Bundles
+	}
+
+	if registryBundles != nil {
```
Comment (timflannagan) [inline pkg/lib/registry/registry.go:357]: Is this the right conditional check? Should we be checking the length of this array instead?

File: pkg/lib/registry/registry.go [hunk 3]
```diff
@@ -342,9 +342,45 @@ func (r RegistryUpdater) DeprecateFromRegistry(request DeprecateFromRegistryRequ
...
+	// Check if all bundlepaths are valid
+	var toDeprecate []string
+
+	dbQuerier := sqlite.NewSQLLiteQuerierFromDb(db)
+	registryBundles, err := dbQuerier.ListBundles(context.TODO())
+	if err != nil {
+		if !request.Permissive {
+			return fmt.Errorf("cannot verify bundle presence in registry: %v", err)
+		}
+		toDeprecate = request.Bundles
```
Comment (timflannagan) [inline pkg/lib/registry/registry.go:354]: Is there any reason we wouldn't want to log that we ran into an error listing the bundles here, even at the debug level? I'm trying to wrap my head around the current permissive workflow, and whether we'd want to ignore if we weren't able to get the list of bundles in the database.

File: pkg/lib/registry/registry.go (hunk 2 above)
Comment (timflannagan) [reply pkg/lib/registry/registry.go:357]: I almost wonder whether this logic would be served being contained in a helper function that we can instead call in this method. It would be easier to add unit tests to hit this case so we don't regress in future releases.

Comment (ankitathomas) [reply pkg/lib/registry/registry.go:357]: It's a edge case we'd almost never run into, but if you try to deprecate a bundle from an empty registry, it wouldn't complain about missing bundles if we look for length.

File: pkg/lib/registry/registry.go [hunk 4]
```diff
@@ -342,9 +342,45 @@ func (r RegistryUpdater) DeprecateFromRegistry(request DeprecateFromRegistryRequ
...
+		for _, bundlePath := range request.Bundles {
+			if _, ok := validPath[bundlePath]; !ok {
+				missingPaths = append(missingPaths, bundlePath)
+				continue
+			}
+			toDeprecate = append(toDeprecate, bundlePath)
+		}
+
+		if len(missingPaths) > 0 {
+			if !request.Permissive {
```
Comment (njhale) [inline pkg/lib/registry/registry.go:374]: IIRC, permissive should decide to return an error at the top-level command, not inform whether an error is returned. I think the proper thing to do in this case is to aggregate all errors and return them at the end of the method.

File: pkg/lib/registry/registry.go [hunk 5]
```diff
@@ -354,3 +368,39 @@ func (r RegistryUpdater) DeprecateFromRegistry(request DeprecateFromRegistryRequ
...
+			if found == nil {
+				found = make([]string, 0)
+			}
+			found = append(found, b)
+			continue
+		}
+		if missing == nil {
+			missing = make([]string, 0)
+		}
+		missing = append(missing, b)
```
Comment (timflannagan) [inline pkg/lib/registry/registry.go:403]: Am I missing something here? It looks like missing was never declared before referencing it here.

File: pkg/lib/registry/registry.go [hunk 6]
```diff
@@ -354,3 +368,39 @@ func (r RegistryUpdater) DeprecateFromRegistry(request DeprecateFromRegistryRequ
...
+	}
+	return
```
Comment (timflannagan) [inline pkg/lib/registry/registry.go:405]: nothing blocking, but this return is redundent.

File: pkg/lib/registry/registry.go (hunk 5 above)
Comment (ankitathomas) [reply pkg/lib/registry/registry.go:403]: missing was declared in the function signature, it looked easier to understand CheckForBundlePaths that way

File: pkg/lib/registry/registry.go (hunk 6 above)
Comment (ankitathomas) [reply pkg/lib/registry/registry.go:405]: it's autopopulated with the args from the signature but the return itself still needs to be here

File: pkg/lib/registry/registry.go (hunk 3 above)
Comment (ankitathomas) [reply pkg/lib/registry/registry.go]: It actually doesn't make much sense to try and remove bundles when you can't even locate them, permissive mode shouldn't even be used here. I changed to fail when those checks fail

File: pkg/lib/registry/registry.go (hunk 6 above)
Comment (joelanford) [reply pkg/lib/registry/registry.go:405]: Could we avoid using named return values in function signatures? 

IMO, unless the function is very short  (like <5 lines), it harms readability.

https://tour.golang.org/basics/7

File: pkg/lib/registry/registry.go [hunk 7]
```diff
@@ -354,3 +373,40 @@ func (r RegistryUpdater) DeprecateFromRegistry(request DeprecateFromRegistryRequ
...
+		if _, ok := registryBundlePaths[b]; ok {
+			if found == nil {
+				found = make([]string, 0)
+			}
+			found = append(found, b)
+			continue
+		}
+		if missing == nil {
+			missing = make([]string, 0)
+		}
```
Comment (njhale) [inline pkg/lib/registry/registry.go:408]: this isn't necessary -- `append` works with `nil` field values so long as the field is declared as a slice type.

File: pkg/lib/registry/registry.go [hunk 8]
```diff
@@ -354,3 +373,40 @@ func (r RegistryUpdater) DeprecateFromRegistry(request DeprecateFromRegistryRequ
...
+	for _, b := range registryBundles {
+		registryBundlePaths[b.BundlePath] = struct{}{}
+	}
+
+	var found, missing []string
+	for _, b := range bundlePaths {
+		if _, ok := registryBundlePaths[b]; ok {
+			if found == nil {
+				found = make([]string, 0)
+			}
```
Comment (njhale) [inline pkg/lib/registry/registry.go:402]: this isn't necessary -- `append` works with `nil` field values so long as the field is declared as a slice type.

File: pkg/lib/registry/registry.go [hunk 9]
```diff
@@ -354,3 +373,40 @@ func (r RegistryUpdater) DeprecateFromRegistry(request DeprecateFromRegistryRequ
...
+	registryBundles, err := querier.ListBundles(context.TODO())
+	if err != nil {
+		return bundlePaths, nil, err
+	}
+
+	if len(registryBundles) == 0 {
+		return nil, bundlePaths, nil
+	}
+
+	registryBundlePaths := make(map[string]struct{})
```
Comment (njhale) [inline pkg/lib/registry/registry.go:392]:
```suggestion
	registryBundlePaths := map[string]struct{}{}
```

File: pkg/lib/registry/registry.go [hunk 10]
```diff
@@ -354,3 +373,40 @@ func (r RegistryUpdater) DeprecateFromRegistry(request DeprecateFromRegistryRequ
 
 	return nil
 }
+
+// CheckForBundlePaths verifies presence of a list of bundle paths in the registry.
+func CheckForBundlePaths(querier registry.Query, bundlePaths []string) ([]string, []string, error) {
```
Comment (njhale) [inline pkg/lib/registry/registry.go:378]: any reason we want this to be exported?

File: pkg/lib/registry/registry_test.go [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+
+			require.NoError(t, store.Migrate(context.TODO()))
+
+			for _, bundle := range tt.bundles {
+				require.NoError(t, store.AddOperatorBundle(bundle))
+			}
+
+			for _, pkg := range tt.pkgs {
+				require.NoError(t, store.AddPackageChannels(pkg))
+			}
```
Comment (njhale) [inline pkg/lib/registry/registry_test.go:166]: Is it feasible for us to mock the querier instead of building a real database?

**Comment by bparees** [comment]
@ankitathomas @njhale who needs to approve this?

**Comment by njhale** [comment]
> who needs to approve this?

meant to approve this earlier.

/approve
